File: tests/test_framing.py
from serial_link.framing import FLAG, ESC, XOR, encode, FrameDecoder


def roundtrip(info, **kw):
    dec = FrameDecoder()
    frames = dec.feed(encode(info, **kw))
    assert len(frames) == 1
    return frames[0]


def test_basic_roundtrip():
    frame = roundtrip(b"\x10hello")
    assert frame.fcs_ok
    assert frame.info == b"\x10hello"
    assert frame.address == 0xFF and frame.control == 0x03


def test_address_and_control_preserved():
    frame = roundtrip(b"\x01\x02", address=0x42, control=0x13)
    assert frame.address == 0x42 and frame.control == 0x13
    assert frame.info == b"\x01\x02"


def test_stuffing_of_special_octets():
    # FLAG and ESC in the payload must be escaped on the wire but survive decode.
    info = bytes([FLAG, ESC, 0x00, 0x7E, 0x7D, 0x41])
    wire = encode(info)
    # No raw FLAG/ESC inside the frame body (between the delimiters).
    assert FLAG not in wire[1:-1]
    assert all(wire[i] != ESC or True for i in range(len(wire)))  # ESC may appear as escape marker
    frame = roundtrip(info)
    assert frame.info == info


def test_control_chars_escaped_by_default_accm():
    info = bytes(range(0x00, 0x20)) + b"\x10"
    wire = encode(info)
    # Every control char (<0x20) in the body should appear escaped, not raw.
    body = wire[1:-1]
    assert FLAG not in body
    frame = roundtrip(info)
    assert frame.info == info


def test_decoder_handles_split_reads():
    wire = encode(b"\x10abc")
    dec = FrameDecoder()
    frames = []
    for byte in wire:  # feed one byte at a time
        frames.extend(dec.feed(bytes([byte])))
    assert len(frames) == 1 and frames[0].info == b"\x10abc"


def test_leading_garbage_ignored():
    wire = b"\x00\x11garbage" + encode(b"\x10ok")
    frames = FrameDecoder().feed(wire)
    assert len(frames) == 1 and frames[0].info == b"\x10ok"


def test_back_to_back_and_shared_flags():
    a = encode(b"\x10one")
    b = encode(b"\x10two")
    # Shared flag: drop the trailing flag of a (it doubles as b's leading flag).
    stream = a[:-1] + b
    frames = FrameDecoder().feed(stream)
    assert [f.info for f in frames] == [b"\x10one", b"\x10two"]


def test_bad_fcs_flagged_not_dropped():
    wire = bytearray(encode(b"\x10payload"))
    # Corrupt a body byte (index 3 is first info byte after FLAG,addr,ctrl).
    wire[4] ^= 0x01
    frames = FrameDecoder().feed(bytes(wire))
    assert len(frames) == 1 and frames[0].fcs_ok is False


def test_runt_frame_dropped():
    # Two flags with <4 bytes between them is idle/runt, not a frame.
    frames = FrameDecoder().feed(bytes([FLAG, 0x01, 0x02, FLAG]))
    assert frames == []

File: tests/test_framing.cpp
// Mirrors tests/test_framing.py — encode/decode round-trips, stuffing, split
// reads, leading garbage, shared flags, bad-FCS flagging, runt drop.
#include <gtest/gtest.h>

#include "serial_link/framing.hpp"

using namespace serial_link;
using namespace serial_link::framing;

namespace {
Frame roundtrip(const Bytes& info, std::uint8_t addr = DEFAULT_ADDRESS,
                std::uint8_t ctrl = DEFAULT_CONTROL) {
    FrameDecoder dec;
    auto frames = dec.feed(encode(info, addr, ctrl));
    EXPECT_EQ(frames.size(), 1u);
    return frames.at(0);
}
Bytes str(const char* s) { return Bytes(s, s + std::char_traits<char>::length(s)); }
}  // namespace

TEST(Framing, BasicRoundtrip) {
    Bytes info = {0x10, 'h', 'e', 'l', 'l', 'o'};
    Frame f = roundtrip(info);
    EXPECT_TRUE(f.fcs_ok);
    EXPECT_EQ(f.info, info);
    EXPECT_EQ(f.address, 0xFF);
    EXPECT_EQ(f.control, 0x03);
}

TEST(Framing, AddressAndControlPreserved) {
    Frame f = roundtrip({0x01, 0x02}, 0x42, 0x13);
    EXPECT_EQ(f.address, 0x42);
    EXPECT_EQ(f.control, 0x13);
    EXPECT_EQ(f.info, (Bytes{0x01, 0x02}));
}

TEST(Framing, StuffingOfSpecialOctets) {
    Bytes info = {FLAG, ESC, 0x00, 0x7E, 0x7D, 0x41};
    Bytes wire = encode(info);
    // No raw FLAG inside the frame body (between the delimiters).
    for (std::size_t i = 1; i + 1 < wire.size(); ++i) EXPECT_NE(wire[i], FLAG);
    EXPECT_EQ(roundtrip(info).info, info);
}

TEST(Framing, ControlCharsEscapedByDefault) {
    Bytes info;
    for (int i = 0; i < 0x20; ++i) info.push_back(static_cast<std::uint8_t>(i));
    info.push_back(0x10);
    Bytes wire = encode(info);
    for (std::size_t i = 1; i + 1 < wire.size(); ++i) EXPECT_NE(wire[i], FLAG);
    EXPECT_EQ(roundtrip(info).info, info);
}

TEST(Framing, DecoderHandlesSplitReads) {
    Bytes wire = encode(str("\x10" "abc"));
    FrameDecoder dec;
    std::vector<Frame> frames;
    for (std::uint8_t b : wire) {
        for (auto& f : dec.feed(Bytes{b})) frames.push_back(f);
    }
    ASSERT_EQ(frames.size(), 1u);
    EXPECT_EQ(frames[0].info, str("\x10" "abc"));
}

TEST(Framing, LeadingGarbageIgnored) {
    Bytes wire = str("\x00\x11garbage");
    Bytes framed = encode(str("\x10ok"));
    wire.insert(wire.end(), framed.begin(), framed.end());
    auto frames = FrameDecoder().feed(wire);
    ASSERT_EQ(frames.size(), 1u);
    EXPECT_EQ(frames[0].info, str("\x10ok"));
}

TEST(Framing, BackToBackAndSharedFlags) {
    Bytes a = encode(str("\x10one"));
    Bytes b = encode(str("\x10two"));
    Bytes stream(a.begin(), a.end() - 1);  // drop a's trailing flag (shared)
    stream.insert(stream.end(), b.begin(), b.end());
    auto frames = FrameDecoder().feed(stream);
    ASSERT_EQ(frames.size(), 2u);
    EXPECT_EQ(frames[0].info, str("\x10one"));
    EXPECT_EQ(frames[1].info, str("\x10two"));
}

TEST(Framing, BadFcsFlaggedNotDropped) {
    Bytes wire = encode(str("\x10payload"));
    wire[4] ^= 0x01;  // corrupt a body byte
    auto frames = FrameDecoder().feed(wire);
    ASSERT_EQ(frames.size(), 1u);
    EXPECT_FALSE(frames[0].fcs_ok);
}

TEST(Framing, RuntFrameDropped) {
    auto frames = FrameDecoder().feed(Bytes{FLAG, 0x01, 0x02, FLAG});
    EXPECT_TRUE(frames.empty());
}
